# 47. 에러 처리

## 47.1 에러 처리의 필요성

> 에러가 발생하지 않는 코드를 작성하는 것은 불가능하다.

> 직접적으로 에러를 발생하지는 않는 예외적인 상황이 발생할 수도 있다.
> 예외적인 상황에 적절하게 대응하지 않으면 에러로 이어질 가능성이 크다.

## 47.2 try ... catch ... finally 문

> 기본적으로 에러 처리를 구현하는 방법은 크게 두 가지가 있다. 예외적인 상황이 발생하면 반환하는 값을 if 문이나 단축 평가 또는 옵셔널 체이닝 연산자를 통해
> 확인해서 처리하는 방법과 에러 처리 코드를 미리 등록해 두고 에러가 발생하면 에러 처리 코드로 점프하도록 하는 방법이 있다.

> try catch finally 문은 두 번쨰 방법이다. 일반적으로 이 방법을 에러 처리라고 한다.

> try 코드 블록이 우선 실행되고 try 코드 블록에 포함된 문 중에서 에러가 발생하면 발생한 에러는 catch 문의 err 변수에 전달되고 catch 코드 블록이 실행된다.

## 47.3 Error 객체

> Error 생성자 함수는 에러 객체를 생성한다. 에러를 상세히 설명하는 에러 메세지를 인수로 전달 가능하다.

```javascript
const error = new Error('new Error');
```

> Error 생성자 함수가 생성한 에러 객체는 message 프로퍼티와 stack 프로퍼티를 갖는다.
> message 프로퍼티의 값은 Error 생성자 함수에 인수로 전달한 에러 메세지이고, stack 프로퍼티의 값은 에러를 발생시킨 콜스택의 호출 정보를 나타내는 문자열
> 디버깅 목적으로 사용 가능하다.

> 자바스크립트는 Error 생성자 함수를 포함해 7가지의 에러 객체를 생성할 수 있는 Error 생성자 함수를 제공한다.

> SyntaxError, ReferenceError, TypeError, RangeError, URIError, EvalError 생성자 함수가 생성한 에러 객체의 프로토타입은 모두 Error.prototpye 을 상속받는다.

### 47.4 throw 문

> Error 생성자 함수로 에러 객체를 생성한다고 에러가 발생하는 것은 아니다. 에러 객체 생성과 에러 발생은 의미가 다르다.

> throw 문의 표현식은 어떤 값이라도 상관없지만 일반적으로 에러 객체를 지정한다. 에러를 던지면 catch 문의 에러 변수가 생성되고 던져진 에러 객체가 할당된다.

## 47.5 에러의 전파

> 에러는 호출자 방향으로 전파된다. 콜 스택의 아래 방향(실행 중인 컨텍스트가 푸시되기 직전에 푸시된 실행 컨텍스트 방향)으로 전파된다.

```javascript
const foo = () => {
    throw Error('foo');
}

const bar = () => {
    foo(); 
}

const baz = () => {
    bar();
}

try {
    baz();
} catch (e) {
    console.error(e);
}
```

> 주의할 점은 비동기 함수인 setTimeout 이나 프로미스 후속 처리 메서드의 콜백 함수는 호출자가 없다는 것이다.
> 콜 스택에 푸시된 콜백 함수의 실행 컨텍스트는 콜 스택의 가장 하부에 존재하게 된다.